sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,365 @@
+	// both to ensure that Run() and Close() are executed at most once each,
+	// and to ensure that the listener cannot be run after it has been closed.
+	// Must be read/modified atomically, and can only be changed in one of the
+	// following ways:
+	// - statusReady -> statusRunning
+	// - statusReady -> statusClosed
+	// - statusRunning -> statusClosed
+	status uint32
+}
+
+const (
+	statusReady uint32 = iota
+	statusRunning
+	statusClosed
+)
+
+// Register opens and configures the apparmor notification interface.
+//
+// If the kernel does not support the notification mechanism the error is ErrNotSupported.
+func Register() (listener *Listener, err error) {
+	path := notify.SysPath
+	if override := os.Getenv("PROMPT_NOTIFY_PATH"); override != "" {
+		path = override
+	}
+
+	notifyFile, err := osOpen(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrListenerNotSupported
+		}
+		return nil, fmt.Errorf("cannot open %q: %v", path, err)
+	}
+	defer func() {
+		if err != nil {
+			notifyFile.Close()
+		}
+	}()
+
+	msg := notify.MsgNotificationFilter{ModeSet: notify.APPARMOR_MODESET_USER}
+	data, err := msg.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	ioctlBuf := notify.IoctlRequestBuffer(data)
+	_, err = notifyIoctl(notifyFile.Fd(), notify.APPARMOR_NOTIF_SET_FILTER, ioctlBuf)
+	if err != nil {
+		return nil, fmt.Errorf("cannot notify ioctl to modeset user on %q: %v", path, err)
+	}
+
+	poll, err := epoll.Open()
+	if err != nil {
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			poll.Close()
+		}
+	}()
+	if err = poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener = &Listener{
+		reqs: make(chan *Request),
+		errs: make(chan error, 1),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+// Close stops the listener and closes the kernel communication file.
+// Returns once all waiting goroutines terminate and the communication file
+// is closed.
+func (l *Listener) Close() error {
+	l.tomb.Kill(nil)
+	origStatus := atomic.SwapUint32(&l.status, statusClosed)
+	switch origStatus {
+	case statusReady:
+		// A goroutine was never spawned with the listener tomb.
+		// Spawn one, so the tomb can die.
+		l.tomb.Go(func() error {
+			<-l.tomb.Dying()
+			return nil
+		})
+	case statusRunning:
+	case statusClosed:
+		return ErrListenerAlreadyClosed
+	}
+	// Stop the run loop from waiting on epoll events
+	err1 := l.poll.Close()
+	// Wait for all waiters to send back deny responses to the kernel
+	l.tomb.Wait()
+	err2 := l.notifyFile.Close()
+	if err1 != nil {
+		return err1
+	}
+	return err2
+}
+
+// Reqs returns a read-only channel through which requests may be received.
+func (l *Listener) Reqs() <-chan *Request {
+	return l.reqs
+}
+
+// Errs returns a read-only channel through which errors may be received.
```

Comment:
this doesn't make ti very clear that at most one error will be sent through errs?